`sage/stats/distributions`: Implement non-spherical Gaussian sampling over lattices

grhkm21 opened the pull request:
I implemented non-spherical Gaussian sampling over lattices. It's useful for cryptography.

~~There are some TODOs that I hope to fix as soon as possible before merging, hence it's in draft status.~~ Done!

There are some missing features e.g. when $B$ is not full-rank, or computing the normalisation factor for non-integral lattices, etc. But those should be easy to fix.

Suggestions to variable naming would be appreciated too :)

TODO:
- [x] Finish writing doctest
- [x] Support taking a basis $B$ and compute $\Sigma = BB^T$ instead.
- [x] Allow setting $c$ dynamically.
- [x] Fix merge conflicts :(
- [x] Fix linting issues, coding style issues, ...

grhkm21 committed:
Compute normalisation const of lattice DGS faster

I rewrote the code for `_normalisation_factor_zz`. It used to enumerate
short vectors, which might be imprecise even for a short basis. The new
method uses `pari.qfrep` to enumerate vectors of a bounded norm + speeds
up convergence of series using Poisson summation.

TODO: The code doesn't work for non-integral lattices (i.e. over QQ),
fixable by rescaling. The code is incorrect for non self-dual lattices,
fixable by implementing the correct formula.

grhkm21 committed:
Compute `_normalisation_factor_zz` for sigma > 0

The formula has been cross-checked with an external C program for a
smaller lattice.

grhkm21 committed:
Implement non-spherical discrete Gaussian sampling

grhkm21 committed:
Add doctest to `._maximal_r()`

grhkm21 committed:
merge

grhkm21 committed:
Refactor code and isolate precomputation

grhkm21 committed:
lint: Fix pycodestyle

grhkm21 committed:
Merge branch 'develop' into pr-lattice

grhkm21 committed:
lint: Fix relint imports

grhkm21 committed:
Fix coverage

grhkm21 committed:
Fix doctest

Why are there so many bugs :/

tscrim commented on src/sage/stats/distributions/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .discrete_gaussian_integer import DGI
+from .discrete_gaussian_lattice import DGL
+from .discrete_gaussian_polynomial import DGP
```
Generally speaking, we keep the `__init__.py` files empty. The `all.py` files handle the imports.

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -279,34 +449,103 @@ def __init__(self, B, sigma=1, c=None, precision=None):
             ....:     add_samples(1000)
 
-            sage: from sage.stats.distributions.discrete_gaussian_lattice import DiscreteGaussianDistributionLatticeSampler
+        Spherical covariance are automatically handled.
```
```suggestion
        Spherical covariance are automatically handled::
```
Also all of the ones below should have this ending as it signifies the next block is code (which is meant to be tested).

tscrim reviewed (changes requested):
Thank you for the contribution. I have a few general comments.

I am generally against having acronyms like `DGI` imported into the global namespace. It clutters the tab completion and the acronym doesn't really describe what it does. There can be exceptions to this, but generally speaking, you should at the very least import the fully spelled out object. I could also see this running into conflicts with something else down the road too. In particular, this is evidenced by the original doctests.

Another is all methods and functions need doctests (with functions defined in functions being an exception, but you should have a fairly decent reason for not splitting it off as a separate method). Similarly, the one-line description should be in the form "Return foo" (not "Returns foo"); this is something we are moving towards universally, so even if other methods do this, let's not make the issue worse.

Any (new) imports to the global namespace should be done with `lazy_import` to make sure the startup time is not increased.

We generally do not have setters (as we generally don't use `@properties`). Even more so, we consider most objects immutable based on their input data. Of course, there are very notable exceptions (matrices, vectors, graphs), but by having things (effectively) immutable, it means that you cannot change things and get unintended side effects.

Lastly,

grhkm21 committed:
Remove @property usage + styles

grhkm21 committed:
Please the lint god ✨

grhkm21 committed:
Merge branch 'develop' into pr-lattice

grhkm21 committed:
Added catalog 📙

tscrim commented on src/sage/stats/all.py:
```diff
@@ -1,7 +1,9 @@
 from .r import ttest
 from .basic_stats import (mean, mode, std, variance, median, moving_average)
 from .hmm import all as hmm
+from .distributions.all import *
```
```suggestion
```
Now that we have the catalog, we can remove the global imports.

tscrim commented on src/sage/stats/distributions/all.py:
```diff
@@ -0,0 +1,6 @@
+# We lazy_import the following modules since they import numpy which
+# slows down sage startup
+from sage.misc.lazy_import import lazy_import
+lazy_import("sage.stats.distributions.discrete_gaussian_integer", ["DiscreteGaussianDistributionIntegerSampler"])
+lazy_import("sage.stats.distributions.discrete_gaussian_lattice", ["DiscreteGaussianDistributionLatticeSampler"])
+lazy_import("sage.stats.distributions.discrete_gaussian_polynomial", ["DiscreteGaussianDistributionPolynomialSampler"])
```
Likewise, we can make this file empty. Although its fine to keep it as-is.

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -14,17 +14,19 @@
-    sage: from sage.stats.distributions.discrete_gaussian_lattice import DiscreteGaussianDistributionLatticeSampler
-    sage: D = DiscreteGaussianDistributionLatticeSampler(ZZ^10, 3.0)
+    sage: from sage.stats.distributions.discrete_gaussian_lattice import DiscreteGaussianDistributionLatticeSampler as DGL
+    sage: D = DGL(ZZ^10, 3.0)
```
It would be good to use the catalog here (and in many other doctests).

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -185,28 +190,30 @@ def compute_precision(precision, sigma):
+          increasing norm until the sum converges to given precision. For high
+          dimension lattice, this is recommended.
+
+        - ``prec`` -- (default: ``None``) Passed to :meth:`compute_precision`
```
```suggestion
        - ``prec`` -- (default: ``None``) passed to :meth:`compute_precision`
```

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -222,51 +229,214 @@ def _normalisation_factor_zz(self, tau=3):
+            raise NotImplementedError("Lattice must contain 0 for now.")
+
+        if self.B.base_ring() != ZZ:
+            raise NotImplementedError("Lattice must be integral for now.")
```
```suggestion
            raise NotImplementedError("only implemented for the basis being a square matrix")

        if self.is_spherical and not self._c_in_lattice:
            raise NotImplementedError("only implemented when the lattice contains 0")

        if self.B.base_ring() != ZZ:
            raise NotImplementedError("only implemented for integral lattices")
```
We also follow Python's convention for error messages (lower case without a period/full-stop).

grhkm21 replied:
Ooo I didn't know it's a Python convention, but I learned that in the past few months

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -222,51 +229,214 @@ def _normalisation_factor_zz(self, tau=3):
+            sage: assert all(e_val >= -1e-12 for e_val in e_vals)
+        """
+        if self.is_spherical:
+            raise RuntimeError("You have encountered a bug. File it! :)")
```
```suggestion
        assert self.is_sperical
```
I am not `00% sure versus the `RuntimeError. I think the assert is slightly better here, but I am not sure I could explain why I feel that way.

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -222,51 +229,214 @@ def _normalisation_factor_zz(self, tau=3):
+        res = (v[0] / (Q * v)[0]).sqrt()
+        return res
+
+    def _randomise(self, v):
```
This method needs a doctest.

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -222,51 +229,214 @@ def _normalisation_factor_zz(self, tau=3):
         - ``precision`` -- bit precision `≥ 53`.
 
+        - ``sigma_basis`` -- (default: False) When set, ``sigma`` is treated as
+            a (row) basis, i.e. the covariance matrix is computed by ``Σ = SSᵀ``.
```
```suggestion
        - ``c`` -- (default: 0) center `c`, any vector in `\ZZ^n` is
          supported, but `c ∈ \Lambda(B)` is faster
        - ``r`` -- (default: ``None``) rounding parameter `r` as defined in
          [Pei2010]_; ignored for spherical Gaussian parameter; if not provided,
          set to be the maximal possible such that `\Sigma - SS^T` is positive
          definite
        - ``precision`` -- bit precision `≥ 53`
        - ``sigma_basis`` -- (default: ``False``) when set, ``sigma`` is treated as
            a (row) basis, i.e. the covariance matrix is computed by ``Σ = SSᵀ``
```

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -317,132 +556,236 @@ def __init__(self, B, sigma=1, c=None, precision=None):
+        .. note::
+
+           Do not call this method directly, it is called automatically from :func:`DiscreteGaussianDistributionLatticeSampler.__init__`.
+        """
```
```suggestion
        .. NOTE::

           Do not call this method directly, it is called automatically from
           :func:`DiscreteGaussianDistributionLatticeSampler.__init__`.
        """
```

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -317,132 +556,236 @@ def __init__(self, B, sigma=1, c=None, precision=None):
         return v
 
-    @property
+    def f(self, x):
```
This needs a doctest.

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -479,7 +822,51 @@ def _call(self):
+
+    def add_offline_samples(self, cnt=1):
+        """
+        Precompute samples from B⁻¹D₁ to be used in :meth:`_call_non_spherical`
```
```suggestion
        Precompute samples from `B^{-1} D_1` to be used in :meth:`_call_non_spherical`.
```

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -479,7 +822,51 @@ def _call(self):
+
+        .. note::
+
+           Do not call this method directly, call :func:`DiscreteGaussianDistributionLatticeSampler.__call__` instead.
```
```suggestion
        .. NOTE::

           Do not call this method directly, call
           :func:`DiscreteGaussianDistributionLatticeSampler.__call__` instead.
```

tscrim reviewed (changes requested):
Sorry for taking so long to continue the review. Thank you for the changes. I am glad to like the catalog approach.

I have a few more specific comments below.

grhkm21 committed:
apply review changes 📖

tscrim commented on src/sage/stats/distributions/catalog.py:
```diff
@@ -0,0 +1,33 @@
+
+To import these names into the global namespace, use::
+
+    sage: from sage.stats.channels_catalog import *
```
This doctest is failing. Looks like the import is not quite correct (older version?).

tscrim commented on src/sage/stats/distributions/catalog.py:
```diff
@@ -0,0 +1,33 @@
+To import these names into the global namespace, use::
+
+    sage: from sage.stats.channels_catalog import *
+
+"""
```
```suggestion
"""
```
We typically don't have the blankline here (at least, it is what we are largely moving towards).

tscrim commented on src/sage/stats/distributions/catalog.py:
```diff
@@ -0,0 +1,33 @@
+from sage.misc.lazy_import import lazy_import as _lazy_import
+_lazy_import("sage.stats.distributions.discrete_gaussian_integer", ["DiscreteGaussianDistributionIntegerSampler"])
+_lazy_import("sage.stats.distributions.discrete_gaussian_lattice", ["DiscreteGaussianDistributionLatticeSampler"])
+_lazy_import("sage.stats.distributions.discrete_gaussian_polynomial", ["DiscreteGaussianDistributionPolynomialSampler"])
```
```suggestion
from sage.misc.lazy_import import lazy_import
lazy_import("sage.stats.distributions.discrete_gaussian_integer", ["DiscreteGaussianDistributionIntegerSampler"])
lazy_import("sage.stats.distributions.discrete_gaussian_lattice", ["DiscreteGaussianDistributionLatticeSampler"])
lazy_import("sage.stats.distributions.discrete_gaussian_polynomial", ["DiscreteGaussianDistributionPolynomialSampler"])

del lazy_import
```
No need for the special handling.

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -1,30 +1,31 @@
 # -*- coding: utf-8 -*-
-r"""
-Discrete Gaussian Samplers over Lattices
+r""" Discrete Gaussian Samplers over Lattices
```
```suggestion
r"""
Discrete Gaussian Samplers over Lattices
```
This is how we do our documentation. All one-lines begin on the next line. (Also, IMO, it is also easier to parse the intent and read it.)

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -1,30 +1,31 @@
 
-    `\exp(-|x-c|_2^2/(2σ²))/(∑_{x ∈ Λ} \exp(-|x|_2^2/(2σ²)))`
+    `\exp(-|x - c|_2^2 / (2\sigma^2)) / (\Sigma_{x \in \Lambda} \exp(-|x|_2^2 /
+    (2\sigma^2)))`
```
```suggestion
.. MATH::

    \frac{\exp(-|x - c|_2^2 / (2\sigma^2))}{\sum_{x \in \Lambda} \exp(-|x|_2^2 /
    (2\sigma^2))}.
```
Also $\sum_{a \in A}$ is different than $\Sigma_{a \in A}$.

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -185,28 +187,29 @@ def compute_precision(precision, sigma):
-        \exp(-|x|_2^2/(2σ²))`, i.e. the normalisation factor such that the sum
-        over all probabilities is 1 for `\ZZⁿ`.
+        This function returns an approximation of `\Sigma_{x \in B}
+        \exp(-|x|_2^2 / (2\sigma²))`, i.e. the normalisation factor such that the sum
```
```suggestion
        This function returns an approximation of `\sum_{x \in B}
        \exp(-|x|_2^2 / (2\sigma^2))`, i.e. the normalization factor such that the sum
```
We also typically use American English spellings (but you don't have to go and fix this everywhere; I just noticed it here).

grhkm21 replied:
Oh no 😂 What should the method be named then? If the American spelling is preferred *and* this is an internal method, might as well change it right?

tscrim replied:
Internal methods can be changed without any deprecation (which is only needed for anything public). Feel free to change it or decide to keep it as-is. (For reference, I am American, so I am biased in this. :P )

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -222,51 +225,218 @@ def _normalisation_factor_zz(self, tau=3):
+    @cached_method
+    def _maximal_r(self):
+        r"""
+        This function computes the largest value `r > 0` such that `Σ - r²BBᵀ`
```
```suggestion
        This function computes the largest value `r > 0` such that `\Sigma - r^2 BB^T`.
```

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -222,51 +225,218 @@ def _normalisation_factor_zz(self, tau=3):
+    def __init__(self, B, sigma=1, c=0, r=None, precision=None, sigma_basis=False):
         r"""
-        Construct a discrete Gaussian sampler over the lattice `Λ(B)`
+        Construct a discrete Gaussian sampler over the lattice `\LAmbda(B)`
```
```suggestion
        Construct a discrete Gaussian sampler over the lattice `\Lambda(B)`
```

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -317,137 +555,250 @@ def __init__(self, B, sigma=1, c=None, precision=None):
+            [0 1 0]
+            [0 0 1]
+
+        .. note:
```
```suggestion
        .. NOTE::
```

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -458,16 +809,16 @@ def _call(self):
+            sage: norm(mean_L.n() - D.c()) < 0.25
             True
 
         .. note::
```
```suggestion
        .. NOTE::
```
A while-we-are-at-it thing.

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -483,3 +834,46 @@ def _call(self):
+
+    def add_offline_samples(self, cnt=1):
+        """
+        Precompute samples from `B^{-1}D_1` to be used in :meth:`_call_non_spherical`
```
```suggestion
        Precompute samples from `B^{-1}D_1` to be used in :meth:`_call_non_spherical`.
```

tscrim reviewed (changes requested):
Thank you. I am just slightly worried about the unicode (within latex markings too) not translating properly into mathjax and/or the pdf docbuild. In principle., it should be fine, but proper latex should always work (provided it is not too fancy). So I am probably just being paranoid, but I definitely know this way will work.

Some other little details that are in the specific comments:

- All one-line descriptions in the docstring start on their own line and end with a period/full-stop except for the module "title".
- We use `.. NOTE::` for all sphinx note formatted blocks.

I think this is just about ready otherwise. Thank you for all the changes and polishing!

grhkm21 committed:
Incorrect docstring formatting 📖

Let's just say I was slightly sleep deprived when I made the previous
commits.

...

grhkm21 committed:
catalog.py fix

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -3,28 +3,32 @@
+.. MATH::
+
+    `\frac{\exp(-|x - c|_2^2 / (2\sigma^2))}{\sum_{v \in \Lambda} \exp(-|v|_2^2 /
+    (2\sigma^2))}`
```
```suggestion
    \frac{\exp(-|x - c|_2^2 / (2\sigma^2))}{\sum_{v \in \Lambda} \exp(-|v|_2^2 /
    (2\sigma^2))}.
```
The extra backticks are not needed here.

tscrim commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -222,51 +228,217 @@ def _normalisation_factor_zz(self, tau=3):
+    @cached_method
+    def _maximal_r(self):
+        r"""
+        This function computes the largest value `r > 0` such that `\Sigma - r^2BB^@`
```
Is the `@` here intentional or should it be a `T`?

grhkm21 replied:
It's a `T`, thanks.

tscrim reviewed (commented):
Sorry for loosing track of this. Thank you for the changes. Just a few small more little details, then I will set a positive review.

grhkm21 committed:
🎉 finally positive review 🎉

grhkm21 committed:
Merge remote-tracking branch 'origin/develop' into pr-lattice

tscrim reviewed (approved):
Thank you.

tornaria commented on src/sage/stats/distributions/discrete_gaussian_lattice.py:
```diff
@@ -279,34 +451,102 @@ def __init__(self, B, sigma=1, c=None, precision=None):
+        The sampler supports non-spherical covariance in the form of a Gram
+        matrix::
+
+            sage: n = 3
+            sage: Sigma = Matrix(ZZ, [[5, -2, 4], [-2, 10, -5], [4, -5, 5]])
+            sage: c = vector(ZZ, [7, 2, 5])
+            sage: D = distributions.DiscreteGaussianDistributionLatticeSampler(ZZ^n, Sigma, c)
+            sage: nf = D._normalisation_factor_zz(); nf # This has not been properly implemented
+            63.76927...
+            sage: while v not in counter: add_samples(1000)
+            sage: while abs(m*f(v)*1.0/nf/counter[v] - 1.0) >= 0.1: add_samples(1000)
```
This test seems wrong... 

Unless we are lucky, this will never finish. Note that `counter` was not reset, we are running the test with the vector from the previous sample, and sometimes the test passes even with a different normalization factor. But if the test doesn't pass on the first few attempts, it won't pass because `m` tends to infinity while `counter[v]` remains constant.

Make sure you reset `counter` and you will see this test looping forever always.

grhkm21 replied:
Sorry about this, I didn’t test with —long much (but why did CI not catch it?)

Can you fix it for me? I am in exam seasons right now, sorry

tornaria replied:
This is not with `--long`. It's a particular seed that causes the failure.

The test passes by chance, from leftover samples of the previous test.

In fact, the normalization factor seems about 20% off, so the test (if properly implemented) would not have much chance of passing:
```
$ sage
┌────────────────────────────────────────────────────────────────────┐
│ SageMath version 10.4.beta3, Release Date: 2024-04-12              │
│ Using Python 3.12.3. Type "help()" for help.                       │
└────────────────────────────────────────────────────────────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Warning: this is a prerelease version, and it may be unstable.     ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
sage: D = distributions.DiscreteGaussianDistributionLatticeSampler(ZZ^3, Matrix(ZZ, [[5, -2, 4], [-2, 10, -5], [4, -5, 5]]), vector(ZZ, [7, 2, 5]))
sage: nf = D._normalisation_factor_zz()
/usr/lib/python3.12/site-packages/sage/stats/distributions/discrete_gaussian_lattice.py:298: UserWarning: Note: `_normalisation_factor_zz` has not been properly implemented for non-spherical distributions.
  warn("Note: `_normalisation_factor_zz` has not been properly "
sage: %time m = 100_000; sum(D() == vector(ZZ,[7,2,5]) for _ in range(m)) * nf / m
CPU times: user 11.5 s, sys: 67.4 ms, total: 11.5 s
Wall time: 11.6 s
0.813695927736100
sage: %time m = 100_000; sum(D() == vector(ZZ,[7,2,5]) for _ in range(m)) * nf / m
CPU times: user 11.3 s, sys: 62.7 ms, total: 11.4 s
Wall time: 11.4 s
0.790738989336022
sage: %time m = 100_000; sum(D() == vector(ZZ,[7,2,5]) for _ in range(m)) * nf / m
CPU times: user 11.5 s, sys: 46.8 ms, total: 11.5 s
Wall time: 11.6 s
0.783086676535996
```

tornaria replied:
Ping @grhkm21 @tscrim

Can we either get this fixed before 10.4 or revert it until it can be fixed properly?

tscrim replied:
Indeed, the test is bad because it does not reset `v` nor `counter`, so it is using the same one from the previous test (which has a similar enough normalization factor). This is why it wasn't getting picked up generally by the tester.

I don't think I would be able to fix this myself. I propose we effectively disable this test with `# known bug`. Reverting the whole PR feels a bit extreme since the normalization doesn't seem to be a key part of the PR (it is hidden, not called from anywhere as far as I can tell, and emits a warning).

grhkm21 replied:
Sorry for the late reply. I fixed it, it's just copying some lines above. Please kindly review #38209 and sorry for the bad code :(
